### ISSUE TITLE
[RFC]: Returning Last Token, Last Layer Hidden States Through The OpenAI API

### ISSUE THREAD

**Issue by kyle-pena-kuzco** [vllm-project/vllm#18176]
### Motivation.

As a distributed inference provider, we use vLLM because it is fast, has cutting edge capabilities, and is well-maintained.  

We serve millions of inference requests per day, all using the OpenAI API format.

Because we run a distributed network, we need to verify that our operators ran the model that was requested (and did not substitute some smaller model or use a canned response).

[TopLOC](https://arxiv.org/pdf/2501.16007) is a very robust inference verification technique.  However, TopLOC requires  the last layer's hidden states for a single token (typically the last token) of the response, something that isn't exposed through the OpenAI API.

Rather than rebuilding an entire API stack, we'd love to leverage vLLM's OpenAI API by incorporating the ability to return the last layer's hidden states, but specifically for the last token only.  Then, we will be able to easily implement TopLOC verification as part of our network.

Please note that this is a different use case from mechaninterp.

### Proposed Change.

We are proposing the following schema changes:

* ServerArgs: Add `return_hidden_states` to turn the feature on at the engine level.

**`ChatCompletionResponseChoice` and `CompletionResponseChoice`:**
_Change:_ Add `hidden_states: Optional[List[List[float]]] = None`, and omit the field entirely when not configured to return it.

**`ChatCompletionResponseStreamChoice` and `CompletionResponseStreamChoice`:**
_Change:_ Add `hidden_states: Optional[List[List[float]]] = None` to DeltaMessage on the final chunk (or possibly the usage chunk), and omit the field entirely when not configured to return it.

3.1-8b has a hidden dimension of 4096, so a hidden state would be of shape [1, 4096].

This is not small, but the payload size is smaller than requesting top 5 logprobs from a response of 1,000 tokens, a feature which the OpenAI API supports (1,000 tokens is not an unusual response size).

### CC List.

@simon-mo 

### Any Other Things.

**We would like to scope the implementation of this RFC to v0, at least initially.**

v1 has architectural constraints that make returning last token, last layer hidden states from the core engine somewhat more challenging.  This requires some careful thought from our team.

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
We've decided against supporting returning hidden states in the general case, see the performance implications that are mentioned in #15434

cc @WoosukKwon @simon-mo in case that has changed

**Comment by simon-mo** [comment]
I asked for this RFC given it's a very specific on last hidden states of last token for a production use case.

**Comment by atbe** [comment]
Thank you @simon-mo, any feedback on the RFC? We'd love to jump into implementing a PR once we're aligned on this.

**Comment by atbe** [comment]
For reference @WoosukKwon the conversation we're having with @simon-mo in Slack can be found here https://vllm-dev.slack.com/archives/C07QP347J4D/p1747266648631689?thread_ts=1747084619.046619&cid=C07QP347J4D Simon has been super helpful so far 👍

**Comment by DarkLight1337** [comment]
cc @maxdebayser since we also have #12249

**Comment by atbe** [comment]
hey @DarkLight1337 @maxdebayser @simon-mo any updates on this RFC?

**Comment by kyle-pena-kuzco** [comment]
@simon-mo Any updates on the RFC?  We were thinking of going for a v0 implementation for starters.

**Comment by atbe** [comment]
> Given we will deprecate V0 soon and the code will be subject to deletion, we prefer all new changes to land in V1.
> 
> There are two questions from the engine design
> 
> * how much memory are we going to use. Does the existing zmq based communication support that optional payload side, and what's the efficient way to serialize the data.
> * performance impact due to the blocking rpc in vLLM. How much does it slow down processing and can this be another DDoS vector.

Hi @simon-mo we actually prefer to run V0 in production and won't mind continuing to run it after it is deprecated. In fact, its probably a better idea for the vLLM team to allow us to merge this feature into V0 considering it won't be supported long term, it will create less burden on the team.

We have already began considering the implementation for V1 and recognize that it will be significantly more complicated to implement because of the new architecture. We would love to contribute a similar feature for V1 at some point in the future.

For now, getting this merged into V0 would be extremely valuable for us, so please consider allowing this change to V0.

Regarding your two questions from the engine design, my colleague @kyle-pena-kuzco will post a response here shortly.

**Comment by kyle-pena-kuzco** [comment]
> Given we will deprecate V0 soon and the code will be subject to deletion, we prefer all new changes to land in V1.
> 
> There are two questions from the engine design
> 
> * how much memory are we going to use. Does the existing zmq based communication support that optional payload side, and what's the efficient way to serialize the data.
> * performance impact due to the blocking rpc in vLLM. How much does it slow down processing and can this be another DDoS vector.

Q: * **how much memory are we going to use.** 
A: We are planning to only transmit the hidden states of the last layer, and only for the last token in a sequence, and only when the feature is enabled.  Assuming a hidden dimension size of 7168 of 16 bit floats, this is an additional 112 kb to serialize per sequence.

Q: * **Does the existing zmq based communication support that optional payload side, and what's the efficient way to serialize the data.**
A: I think you are asking if it is possible to serialize a field of type `Optional[np.ndarray]` or `Optional[list[float]]` (let me know if I have misinterpreted the question)?  

From what I see, Optional fields are supported. I would need to instrument to determine if `list[float]` or `np.ndarray` is more efficient for the typical sizes involved.  There are some comments in vLLM suggesting that `list[float]` is preferred over a `np` or `torch` tensor.

Q: **performance impact due to the blocking rpc in vLLM. How much does it slow down processing**
- This feature would only be active if turned on at the engine level (i.e.;  `--enable-return-hidden-states`), and would be implemented in a way such that no performance cost would be incurred if it is turned off.
- Speculative decoding already captures and utilizes the same information (last layer hidden states).  If there is any additional performance impact, I think it would be from either:
   (a) Moving a single torch slice of shape (D,) (i.e.; 7,168) to the CPU and converting it to a list of floats, once per sequence
   (b) serializing and deserializing this `list[float]` over zmq.

Q:  **can this be another DDoS vector.**
It's hard to say without a complete PR, but I don't believe so. 
Because:
  (a) the feature has to be specifically enabled, meaning any risk is isolated to servers that have it turned on
  (b) we are using information that is already available to speculative decoding
  (c) i expect total additional serialization time per sequence to be comparable to existing fields like `LogProbLists`.

**Comment by kyle-pena-kuzco** [comment]
Hi @simon-mo - thanks again for your consideration.

TLDR; I don't anticipate to consume more memory than is already used for spec decoding, and I think we will add an additional ~50-100kb of data to serialize per sequence when the feature is turned on at the engine level, with no cost in memory or serialization when the feature is turned off.

**Comment by kyle-pena-kuzco** [comment]
@simon-mo 

We have given a v1 implementation careful thought and have built a proof of concept.

Here is our proposal:

* Add a flag `request_hidden_states` in vllm/sampling_params.py
* If hidden states are requested for a sequence, when the sequence is complete, schedule a additional, final request for extracting hidden states (`HiddenStatesExtractionRequest`).
* A `HiddenStatesExtractionRequest` specifies which token index to extract hidden states for.  For example, `-1` (the default) indicates the last token.
* The core handles a `HiddenStatesExtractionRequest` by performing a prefill of prompt_input_ids=prompt+response, and extracting the hidden states for the requested token index (-1).  That way, getting the hidden states for the last token of the response is fast because it is a prefill operation for the model runner.

We are also adding the glue code to integrate this with the API layer.

This design has the following advantages:
1. No additional cost or overhead for requests that do not request hidden states (no `HiddenStatesExtractionRequest` performed).
2. Preserves the clean separation of concerns between the scheduler and the core engine.
3. Does not require temporary storage of hidden states over the request lifecycle because the hidden states are reconstructed by prefill of prompt+response after the sequence is complete.
4. The default behavior is "last token only" (token index of `-1`).  The total serialization cost is comparable to including `top_logprobs`.

We are doing extensive testing locally to ensure this implementation is compatible with spec decoding, CUDA graph, and KV cache.

**Comment by WoosukKwon** [comment]
@kyle-pena-kuzco 
What do you exactly mean by the "last token"? Do you mean the last token of prefill?
Also, can we implement hashing inside vllm and return hash (instead of the tensor)?

**Comment by kyle-pena-kuzco** [comment]
Hey @WoosukKwon - thank you for the response.  We want to make sure that what we are proposing is acceptable to you and we appreciate the engagement.

> What do you exactly mean by the "last token"? Do you mean the last token of prefill?

If the prompt is: `The capital of France is` and the response is `Paris.<|EOS|>`, 
then the last token would be the special end-of-sequence token `<|EOS|>`.

>  Also, can we implement hashing inside vllm and return hash (instead of the tensor)?

I am concerned about the idea of incorporating TopLOC hashing directly into the vLLM codebase.

TopLOC hashing has settings that are adjusted on a per-model basis.  So vLLM would need to add these TopLOC-specific settings in order to support TopLOC across several models.

Also, TopLOC has two parts: Verification and hashing.  If we don't return hidden states and move the hashing into vLLM, then we would also have to incorporate the verification code into vLLM as well.  That's because verification compares a hash to a hidden states vector, not a hash to a hash.

Lastly, while TopLOC is current SOTA for hidden state hashing and verification, it is possible this may change in the future.  If hidden states are returned by vLLM, this allows us to adopt a new hashing algorithm without submitting another PR for support of another hashing algorithm.

**Comment by WoosukKwon** [comment]
@kyle-pena-kuzco Thanks for the clarification! I still think this adds too much complexity for the main vLLM repo.
From a system design perspective, we’d prefer to keep it simple—ideally by hashing the last hidden states (or logits) within vLLM and sending that to users.

**Comment by kyle-pena-kuzco** [comment]
Thank you for your feedback @WoosukKwon 

I'll see if i can put together an implementation which hashes inside vLLM, and transmits only the hash in the `EngineCoreOutput`.

I'll implement as much of the hashing operation on the GPU as possible in order to avoid GPU->CPU costs.   Because only the hash is included on the EngineCoreOutput, the ZMQ serialization costs should be relatively low.

And no hashing will be performed if the feature is disabled at the server level.

Does this sound like a more acceptable design?

**Comment by WoosukKwon** [comment]
@kyle-pena-kuzco Yes. Thanks for understanding!

**Comment by kyle-pena-kuzco** [comment]
> [@kyle-pena-kuzco](https://github.com/kyle-pena-kuzco) Yes. Thanks for understanding!

Thank you as well!  We appreciate the candid feedback.

**Comment by ywang96** [comment]
> > [@kyle-pena-kuzco](https://github.com/kyle-pena-kuzco) Yes. Thanks for understanding!
> 
> Thank you as well! We appreciate the candid feedback.

@kyle-pena-kuzco Please let us know if you're still interested in pursuing this RFC! If we don't hear back in the next 5 days this RFC will be closed, but feel free to open a new one with your updated plan. Thanks!
